### PR TITLE
Use an array instead of a string for the FQDN parameter

### PR DESCRIPTION
**Pull request by TheBigLee**
## Summary

* This PR will change the FQD parameter from a string to an array, so that the end user can specify multiple endpoints.

## Checklist

- [x] Categorize the PR by setting a good title and adding one of the labels:
      `bug`, `enhancement`, `documentation`, `change`, `breaking`, `dependency`
      as they show up in the changelog
- [x] Update tests.
- [ ] Link this PR to related issues.

### CODE REVIEW

File: apis/vshn/v1/vshn_nextcloud.go [hunk 1]
```diff
@@ -91,11 +91,12 @@ type VSHNNextcloudServiceSpec struct {
...
 	Collabora CollaboraSpec `json:"collabora,omitempty"`
 	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:MinItems=1
 
-	// FQDN contains the FQDN which will be used for the ingress.
+	// FQDN contains the FQDNs which will be used for the ingress.
 	// If it's not set, no ingress will be deployed.
 	// This also enables strict hostname checking for this FQDN.
-	FQDN string `json:"fqdn"`
+	FQDN []string `json:"fqdn"`
```
Comment (wejdross) [inline apis/vshn/v1/vshn_nextcloud.go:99]: remember to inform customer before deployment :P

File: apis/vshn/v1/vshn_nextcloud.go [hunk 2]
```diff
@@ -91,11 +91,12 @@ type VSHNNextcloudServiceSpec struct {
 	// Collabora contains settings to control the Collabora integration.
 	Collabora CollaboraSpec `json:"collabora,omitempty"`
 	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:MinItems=1
 
-	// FQDN contains the FQDN which will be used for the ingress.
+	// FQDN contains the FQDNs which will be used for the ingress.
```
Comment (wejdross) [inline apis/vshn/v1/vshn_nextcloud.go:96]:
```suggestion
	// FQDN contains the FQDNs array, which will be used for the ingress.
```

`make generate` afterwards

File: pkg/comp-functions/functions/vshnnextcloud/deploy.go [hunk 3]
```diff
@@ -256,10 +256,20 @@ func newValues(ctx context.Context, svc *runtime.ServiceRuntime, comp *vshnv1.VS
...
 	}
 
+	if len(comp.Spec.Parameters.Service.FQDN) == 0 {
+		return nil, fmt.Errorf("FQDN array is empty, but requires at least one entry: %w", errors.New("empty fqdn"))
+	}
+
+	trustedDomain := []string{
+		comp.GetName() + "." + comp.GetInstanceNamespace() + ".svc.cluster.local",
+	}
+	trustedDomain = append(trustedDomain, comp.Spec.Parameters.Service.FQDN...)
```
Comment (wejdross) [inline pkg/comp-functions/functions/vshnnextcloud/deploy.go:266]: I'd make here verification, similar to: https://github.com/vshn/appcat/blob/master/pkg/comp-functions/functions/vshnnextcloud/collabora.go#L56

or even now I'm thinking that we could move this input verification to webhook, so we won't receive ticket to debug broken fqdn

**Review by wejdross — CHANGES_REQUESTED** [review]

File: pkg/comp-functions/functions/vshnnextcloud/deploy.go (hunk 3 above)
Comment (TheBigLee) [reply pkg/comp-functions/functions/vshnnextcloud/deploy.go:271]: I've added the validation

**Review by wejdross — APPROVED** [review]

**Comment by wejdross** [comment]
please fix Summary section before merging

File: pkg/comp-functions/functions/vshnnextcloud/deploy.go [hunk 4]
```diff
@@ -428,3 +443,12 @@ func setBackgroundJobMaintenance(t vshnv1.TimeOfDay, nextcloudConfig string) str
 	backgroundJobHour := t.GetTime().Add(40 * time.Minute).Add(time.Hour).Hour()
 	return strings.Replace(nextcloudConfig, "%maintenance_value%", strconv.Itoa(backgroundJobHour), 1)
 }
+
+func validateFQDNs(fqdns []string) error {
+	for _, fqdn := range fqdns {
+		if !valid.IsDNSName(fqdn) {
```
Comment (Kidswiss) [inline pkg/comp-functions/functions/vshnnextcloud/deploy.go:449]: Idea: this is something we could check in a webhook, so it doesn't break silently for the user.

**Review by Kidswiss — APPROVED** [review]
Just one idea, rest LGTM

File: pkg/comp-functions/functions/vshnnextcloud/deploy.go (hunk 4 above)
Comment (TheBigLee) [reply pkg/comp-functions/functions/vshnnextcloud/deploy.go:445]: I've refactored the code and moved this validation to the webhook

File: pkg/controller/webhooks/nextcloud.go [hunk 5]
```diff
@@ -41,3 +47,36 @@ func SetupNextcloudWebhookHandlerWithManager(mgr ctrl.Manager, withQuota bool) e
 		}).
 		Complete()
 }
+
+func (n *NextcloudWebhookHandler) ValidateCreate(ctx context.Context, obj runtime.Object) (admission.Warnings, error) {
```
Comment (Kidswiss) [inline pkg/controller/webhooks/nextcloud.go:51]: We should also validate on update, not only on create.

**Review by Kidswiss — CHANGES_REQUESTED** [review]

**Review by wejdross — APPROVED** [review]

File: pkg/controller/webhooks/nextcloud.go (hunk 5 above)
Comment (TheBigLee) [reply pkg/controller/webhooks/nextcloud.go:51]: Good point. Added that part as well

File: pkg/comp-functions/functions/vshnnextcloud/collabora_test.go [hunk 6]
```diff
@@ -56,11 +56,4 @@ func Test_addCollabora(t *testing.T) {
 	res = DeployCollabora(ctx, comp, svc)
 
 	assert.True(t, res.Severity == v1beta1.Severity_SEVERITY_NORMAL && res.Message == "Collabora not enabled")
-
```
Comment (zugao) [inline pkg/comp-functions/functions/vshnnextcloud/collabora_test.go:59]: well I would add a test here check on the list of FQDN.

Comment (zugao) [reply pkg/comp-functions/functions/vshnnextcloud/collabora_test.go:59]: Why this had to be removed? don't you work with nextcloud fqdn?

File: pkg/comp-functions/functions/vshnnextcloud/deploy.go [hunk 7]
```diff
@@ -256,10 +257,20 @@ func newValues(ctx context.Context, svc *runtime.ServiceRuntime, comp *vshnv1.VS
 		}
 	}
 
+	if len(comp.Spec.Parameters.Service.FQDN) == 0 {
```
Comment (zugao) [inline pkg/comp-functions/functions/vshnnextcloud/deploy.go:260]: Can we add that into directly in the the webhook?

Comment (TheBigLee) [reply pkg/comp-functions/functions/vshnnextcloud/deploy.go]: Done

Comment (Kidswiss) [reply pkg/comp-functions/functions/vshnnextcloud/deploy.go]: > Can we add that into directly in the the webhook?

Already covered here: https://github.com/vshn/appcat/pull/269/files#diff-d41d9990ae83874d678984536634bbcc08146417526791f235c09d1bdc9f0d7eR94

No need for a webhook here...

**Review by Kidswiss — APPROVED** [review]

File: pkg/comp-functions/functions/vshnnextcloud/deploy.go (hunk 7 above)
Comment (TheBigLee) [reply pkg/comp-functions/functions/vshnnextcloud/deploy.go]: True, I've removed the whole section

**Pull request merged**